File: projects/Oxygen.Editor.World/src/LightValidation.cs
// Distributed under the MIT License. See accompanying file LICENSE or copy
// at https://opensource.org/licenses/MIT.
// SPDX-License-Identifier: MIT

using System.Numerics;
using Oxygen.Editor.World.Serialization;

namespace Oxygen.Editor.World;

/// <summary>Validates complete authored light values before hydration, edits or cooking.</summary>
public static class LightValidation
{
    private const double MinimumNormal = 1.1754943508222875e-38;

    /// <summary>Returns the rejected field and reason, or null for a valid candidate.</summary>
    public static string? Validate(LightComponentData light)
    {
        if (!Nonnegative(light.Color)) return "Color must be finite and nonnegative.";
        if (!float.IsFinite(light.ExposureCompensation)) return "Exposure compensation must be finite.";
        var shadow = light.Shadow ?? new();
        if (!Nonnegative(shadow.Bias) || !Nonnegative(shadow.NormalBias)) return "Shadow biases must be finite and nonnegative.";
        if (!Enum.IsDefined(shadow.ResolutionHint)) return "Unknown shadow resolution hint.";
        double solidAngle = 1;
        float strength;
        if (light is DirectionalLightData directional)
        {
            strength = directional.IntensityLux;
            if (!Enum.IsDefined(directional.AtmosphereSlot)) return "Unknown atmosphere light slot.";
            if (!Nonnegative(directional.AngularSizeRadians) || directional.AngularSizeRadians > MathF.PI) return "Angular diameter must be in [0, pi].";
            if (!Nonnegative(directional.AtmosphereDiskLuminanceScaleRgb)) return "Disk scale must be finite and nonnegative.";
            if (directional.CascadeCount is < 1 or > 4 || !Enum.IsDefined(directional.SplitMode)) return "Invalid cascade count or split mode.";
            if (!Nonnegative(directional.MaxShadowDistance) || directional.MaxShadowDistance == 0) return "Shadow distance must be positive.";
            if (!float.IsFinite(directional.DistributionExponent) || directional.DistributionExponent < 1) return "Distribution exponent must be at least one.";
            if (!Fraction(directional.TransitionFraction) || !Fraction(directional.DistanceFadeoutFraction)) return "Shadow fade fractions must be in [0, 1].";
            float previous = 0;
            for (var i = 0; i < 4; ++i)
            {
                var distance = directional.CascadeDistances[i];
                if (!float.IsFinite(distance) || (i < directional.CascadeCount && distance <= previous)) return "Active cascade distances must be positive and increasing; inactive entries must be finite.";
                previous = distance;
            }
        }
        else
        {
            var (range, radius, flux) = light switch
            {
                PointLightData point => (point.Range, point.SourceRadius, point.LuminousFluxLumens),
                SpotLightData spot => (spot.Range, spot.SourceRadius, spot.LuminousFluxLumens),
                _ => (float.NaN, 0f, 0f),
            };
            if (!Nonnegative(range) || !Nonnegative(radius) || (double)range + radius > float.MaxValue) return "Local range and radius must be finite and nonnegative.";
            if (range > 0 && (1.0 / range < MinimumNormal || 1.0 / range > float.MaxValue)) return "Inverse light range is not representable.";
            strength = flux;
            solidAngle = 4 * Math.PI;
            if (light is SpotLightData spotLight)
            {
                var inner = spotLight.InnerConeAngleRadians;
                var outer = spotLight.OuterConeAngleRadians;
                if (!Nonnegative(inner) || !float.IsFinite(outer) || outer <= 0 || inner > outer || outer > MathF.PI / 2 || (inner == outer && outer == MathF.PI / 2)) return "Invalid spot cone pair.";
                var innerSineSquared = inner == MathF.PI / 2 ? 0.5 : Math.Pow(Math.Sin(inner / 2.0), 2);
                var outerSineSquared = outer == MathF.PI / 2 ? 0.5 : Math.Pow(Math.Sin(outer / 2.0), 2);
                var innerCosine = (float)(1 - 2 * innerSineSquared);
                var outerCosine = (float)(1 - 2 * outerSineSquared);
                if (outerCosine >= 1 || (inner != outer && innerCosine <= outerCosine)) return "Spot cone is not representable by the renderer.";
                solidAngle *= innerSineSquared + ((outerSineSquared - innerSineSquared) / 3);
            }
        }

        if (!Nonnegative(strength)) return "Light intensity must be finite and nonnegative.";
        for (var channel = 0; channel < 3; ++channel)
        {
            if (strength == 0 || light.Color[channel] == 0) continue;
            var logIntensity = Math.Log2(strength * (double)light.Color[channel] / solidAngle) + light.ExposureCompensation;
            if (!double.IsFinite(logIntensity) || logIntensity < -126 || logIntensity > Math.Log2(float.MaxValue)) return "Resolved light intensity is not representable.";
            if (light is DirectionalLightData disk && disk.AngularSizeRadians > 0)
            {
                var sine = Math.Sin(disk.AngularSizeRadians / 2.0);
                var radiance = Math.Pow(2, logIntensity) * disk.AtmosphereDiskLuminanceScaleRgb[channel] / (Math.PI * sine * sine);
                if (!double.IsFinite(radiance) || radiance > float.MaxValue || (radiance > 0 && radiance < MinimumNormal)) return "Resolved disk radiance is not representable.";
            }
        }
        return null;
    }

    /// <summary>Validates light ownership and values across stored nodes, including inactive ones.</summary>
    public static string? ValidateScene(Scene scene, IReadOnlyDictionary<Guid, DirectionalLightData>? candidates = null)
    {
        var owners = new Dictionary<AtmosphereLightSlot, string>();
        foreach (var node in scene.AllNodes)
        {
            var lights = node.Components.OfType<LightComponent>().ToArray();
            if (lights.Length > 1) return $"'{node.Name}' has more than one light.";
            if (lights.Length == 0) continue;
            var light = candidates is not null && candidates.TryGetValue(node.Id, out var candidate)
                ? candidate : (LightComponentData)lights[0].Dehydrate();
            if (Validate(light) is { } error) return $"'{node.Name}': {error}";
            if (light is DirectionalLightData { AtmosphereSlot: not AtmosphereLightSlot.None } directional
                && !owners.TryAdd(directional.AtmosphereSlot, node.Name))
            {
                return $"Atmosphere slot {directional.AtmosphereSlot} is already owned by '{owners[directional.AtmosphereSlot]}'.";
            }
        }
        return null;
    }

    /// <summary>Validates stored lights before replacing a scene during hydration.</summary>
    public static string? ValidateScene(SceneData scene)
    {
        var owners = new Dictionary<AtmosphereLightSlot, string>();
        var pending = new Stack<SceneNodeData>(scene.RootNodes.Reverse());
        while (pending.TryPop(out var node))
        {
            var lights = node.Components.OfType<LightComponentData>().ToArray();
            if (lights.Length > 1) return $"'{node.Name}' has more than one light.";
            foreach (var light in lights)
            {
                if (Validate(light) is { } error) return $"'{node.Name}': {error}";
                if (light is DirectionalLightData { AtmosphereSlot: not AtmosphereLightSlot.None } directional
                    && !owners.TryAdd(directional.AtmosphereSlot, node.Name))
                {
                    return $"Atmosphere slot {directional.AtmosphereSlot} is already owned by '{owners[directional.AtmosphereSlot]}'.";
                }
            }
            if (node.Children is { } children)
            {
                foreach (var child in children.Reverse()) pending.Push(child);
            }
        }
        return null;
    }

    private static bool Nonnegative(float value) => float.IsFinite(value) && value >= 0;
    private static bool Nonnegative(Vector3 value) => Nonnegative(value.X) && Nonnegative(value.Y) && Nonnegative(value.Z);
    private static bool Fraction(float value) => Nonnegative(value) && value <= 1;
}

File: projects/Oxygen.Engine/src/Oxygen/Scene/Light/LightValidation.cpp
//===----------------------------------------------------------------------===//
// Distributed under the 3-Clause BSD License. See accompanying file LICENSE or
// copy at https://opensource.org/licenses/BSD-3-Clause.
// SPDX-License-Identifier: BSD-3-Clause
//===----------------------------------------------------------------------===//

#include <cmath>
#include <limits>
#include <numbers>
#include <string_view>

#include <Oxygen/Core/Lighting/LightPhotometry.h>
#include <Oxygen/Scene/Light/DirectionalLight.h>
#include <Oxygen/Scene/Light/LightValidation.h>
#include <Oxygen/Scene/Light/PointLight.h>
#include <Oxygen/Scene/Light/SpotLight.h>

namespace oxygen::scene {
namespace {

  auto Reject(std::string_view field, std::string_view message)
    -> std::optional<LightValidationError>
  {
    return LightValidationError { .field = std::string(field),
      .message = std::string(message) };
  }

  auto Nonnegative(float value) -> bool
  {
    return std::isfinite(value) && value >= 0.0F;
  }

  auto ValidateCommon(const CommonLightProperties& common)
    -> std::optional<LightValidationError>
  {
    if (!Nonnegative(common.color_rgb.x) || !Nonnegative(common.color_rgb.y)
      || !Nonnegative(common.color_rgb.z)) {
      return Reject("color_rgb", "Color components must be finite and nonnegative");
    }
    if (!std::isfinite(common.exposure_compensation_ev)) {
      return Reject("exposure_compensation_ev", "Exposure compensation must be finite");
    }
    if (common.mobility != LightMobility::kRealtime) {
      return Reject("mobility", "Only realtime light authoring is supported");
    }
    if (!Nonnegative(common.shadow.bias)) {
      return Reject("shadow.bias", "Shadow bias must be finite and nonnegative");
    }
    if (!Nonnegative(common.shadow.normal_bias)) {
      return Reject("shadow.normal_bias", "Normal bias must be finite and nonnegative");
    }
    if (common.shadow.resolution_hint > ShadowResolutionHint::kUltra) {
      return Reject("shadow.resolution_hint", "Unknown shadow resolution hint");
    }
    return {};
  }

  auto Modifiers(const CommonLightProperties& common) -> lighting::LightPhotometryModifiers
  {
    return { .color_rgb = common.color_rgb,
      .exposure_compensation_ev = common.exposure_compensation_ev };
  }

  auto ValidateLocal(float range, float source_radius)
    -> std::optional<LightValidationError>
  {
    if (!Nonnegative(range)) {
      return Reject("range", "Range must be finite and nonnegative");
    }
    const auto inverse = range == 0.0F ? 0.0 : 1.0 / double(range);
    if (inverse > std::numeric_limits<float>::max()
      || (inverse > 0.0 && inverse < std::numeric_limits<float>::min())) {
      return Reject("range", "Inverse range is not representable by the renderer");
    }
    if (!Nonnegative(source_radius)
      || double(range) + source_radius > std::numeric_limits<float>::max()) {
      return Reject("source_radius", "Source radius must be finite, nonnegative and representable");
    }
    return {};
  }

  auto ValidateDirectional(const DirectionalLight& light)
    -> std::optional<LightValidationError>
  {
    if (const auto error = ValidateCommon(light.Common())) return error;
    const auto rgb = lighting::ResolveDirectionalIlluminanceRgb(
      light.GetIntensityLux(), Modifiers(light.Common()));
    if (!rgb) {
      return Reject("intensity_lux", "Resolved RGB illuminance must be finite, nonnegative and representable");
    }
    const auto diameter = light.GetAngularSizeRadians();
    if (!Nonnegative(diameter) || diameter > std::numbers::pi_v<float>) {
      return Reject("angular_size_radians", "Angular diameter must be in [0, pi]");
    }
    if (light.GetAtmosphereLightSlot() > AtmosphereLightSlot::kSecondary) {
      return Reject("atmosphere_light_slot", "Unknown atmosphere light slot");
    }
    const auto scale = light.GetAtmosphereDiskLuminanceScale();
    const auto sine = std::sin(double(diameter) * 0.5);
    const auto projected_solid_angle = std::numbers::pi * sine * sine;
    for (int channel = 0; channel < 3; ++channel) {
      if (!Nonnegative(scale[channel])) {
        return Reject("atmosphere_disk_luminance_scale_rgb", "Disk scale must be finite and nonnegative");
      }
      if (diameter == 0.0F) continue;
      const auto radiance = double((*rgb)[channel]) * scale[channel] / projected_solid_angle;
      if (!std::isfinite(radiance) || radiance > std::numeric_limits<float>::max()
        || (radiance > 0.0 && radiance < std::numeric_limits<float>::min())) {
        return Reject("atmosphere_disk_luminance_scale_rgb", "Resolved disk radiance is not representable");
      }
    }
    const auto& csm = light.CascadedShadows();
    if (csm.cascade_count == 0U || csm.cascade_count > kMaxShadowCascades) {
      return Reject("cascade_count", "Cascade count must be in [1, 4]");
    }
    if (!IsValidDirectionalCsmSplitMode(csm.split_mode)) {
      return Reject("split_mode", "Unknown cascade split mode");
    }
    if (!std::isfinite(csm.max_shadow_distance) || csm.max_shadow_distance <= 0.0F) {
      return Reject("max_shadow_distance", "Shadow distance must be finite and positive");
    }
    float previous = 0.0F;
    for (std::size_t i = 0U; i < csm.cascade_distances.size(); ++i) {
      const auto distance = csm.cascade_distances[i];
      if (!std::isfinite(distance) || (i < csm.cascade_count && distance <= previous)) {
        return Reject("cascade_distances", "Distances must be finite; active splits must be positive and increasing");
      }
      previous = distance;
    }
    if (!std::isfinite(csm.distribution_exponent) || csm.distribution_exponent < 1.0F) {
      return Reject("distribution_exponent", "Distribution exponent must be finite and at least one");
    }
    if (!Nonnegative(csm.transition_fraction) || csm.transition_fraction > 1.0F) {
      return Reject("transition_fraction", "Transition fraction must be in [0, 1]");
    }
    if (!Nonnegative(csm.distance_fadeout_fraction) || csm.distance_fadeout_fraction > 1.0F) {
      return Reject("distance_fadeout_fraction", "Fadeout fraction must be in [0, 1]");
    }
    return {};
  }
} // namespace

auto ValidateLight(const Component& light) -> std::optional<LightValidationError>
{
  if (light.GetTypeId() == DirectionalLight::ClassTypeId()) {
    return ValidateDirectional(static_cast<const DirectionalLight&>(light));
  }
  if (light.GetTypeId() == PointLight::ClassTypeId()) {
    const auto& point = static_cast<const PointLight&>(light);
    if (const auto error = ValidateCommon(point.Common())) return error;
    if (const auto error = ValidateLocal(point.GetRange(), point.GetSourceRadius())) return error;
    if (!lighting::ResolvePointIntensityRgb(point.GetLuminousFluxLm(), Modifiers(point.Common()))) {
      return Reject("luminous_flux_lm", "Resolved RGB candela must be finite, nonnegative and representable");
    }
    return {};
  }
  if (light.GetTypeId() == SpotLight::ClassTypeId()) {
    const auto& spot = static_cast<const SpotLight&>(light);
    if (const auto error = ValidateCommon(spot.Common())) return error;
    if (const auto error = ValidateLocal(spot.GetRange(), spot.GetSourceRadius())) return error;
    const auto cone = lighting::ResolveSpotConeProfile(
      spot.GetInnerConeAngleRadians(), spot.GetOuterConeAngleRadians());
    if (!cone) {
      return Reject("cone_angles", "Cone pair must be ordered, within a hemisphere and representable");
    }
    if (!lighting::ResolveSpotIntensityRgb(spot.GetLuminousFluxLm(), *cone, Modifiers(spot.Common()))) {
      return Reject("luminous_flux_lm", "Resolved RGB candela must be finite, nonnegative and representable");
    }
    return {};
  }
  return Reject("type", "Unsupported light component type");
}

} // namespace oxygen::scene
